src/commands/backup.rs

```diff
@@ -206,82 +211,134 @@ impl BackupCmd {
-                    }
-                }
-            }
 
-            // merge "backup" section from config file, if given
-            opts.merge(config.backup.clone());
-
-            let backup_opts = BackupOptions::default()
-                .stdin_filename(opts.stdin_filename)
-                .as_path(opts.as_path)
-                .parent_opts(opts.parent_opts)
-                .ignore_save_opts(opts.ignore_save_opts)
-                .ignore_filter_opts(opts.ignore_filter_opts)
-                .no_scan(opts.no_scan)
-                .dry_run(config.global.dry_run);
-            let snap = repo.backup(&backup_opts, &source, opts.snap_opts.to_snapshot()?)?;
-
-            if opts.json {
-                let mut stdout = std::io::stdout();
-                serde_json::to_writer_pretty(&mut stdout, &snap)?;
-            } else if opts.long {
-                let mut table = table();
-
-                let add_entry = |title: &str, value: String| {
-                    _ = table.add_row([bold_cell(title), Cell::new(value)]);
-                };
-                fill_table(&snap, add_entry);
-
-                println!("{table}");
-            } else if !opts.quiet {
-                let summary = snap.summary.unwrap();
-                println!(
-                    "Files:       {} new, {} changed, {} unchanged",
-                    summary.files_new, summary.files_changed, summary.files_unmodified
-                );
-                println!(
-                    "Dirs:        {} new, {} changed, {} unchanged",
-                    summary.dirs_new, summary.dirs_changed, summary.dirs_unmodified
-                );
-                debug!("Data Blobs:  {} new", summary.data_blobs);
-                debug!("Tree Blobs:  {} new", summary.tree_blobs);
-                println!(
-                    "Added to the repo: {} (raw: {})",
-                    bytes_size_to_string(summary.data_added_packed),
-                    bytes_size_to_string(summary.data_added)
-                );
+            if let Err(err) = backup_source(source.clone(), opts, &repo) {
+                error!("error backing up {source}: {err}");
+                is_err = true;
+            }
+        }
 
-                println!(
-                    "processed {} files, {}",
-                    summary.total_files_processed,
-                    bytes_size_to_string(summary.total_bytes_processed)
-                );
-                println!("snapshot {} successfully saved.", snap.id);
+        if has_sources {
+            if is_err {
+                hooks.run_failed()?;
+            } else {
+                hooks.run_after()?;
             }
+        }
+        hooks.run_finally()?;
 
-            info!("backup of {source} done.");
+        if is_err {
+            bail!("Not all sources were successfully backuped!")
         }
 
         Ok(())
     }
 }
+
+fn backup_source<P: ProgressBars, S: IndexedIds>(
```

Comment:
couldn't/shouldn't this be part of an `impl BackupCmd` ?

something ~close to that?
```rust
impl BackupCmd {
   fn backup_source<P: Pr.., S: Inde..(&mut self, source: PathList, repo: &Repository<P, S>) 
}
```

Reply:
is changed.